src/eval.c

```diff
@@ -5394,6 +5465,774 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    size_t		all_embedding_size = sum_embedding_strlen(embedded, embedded_size);
+    size_t		all_stringified_size = sum_embedded_stringified_strlen(embedded, embedded_size);
+    size_t		template_length = strlen_template(template, is_closing_quote);
+    // NOTE: e.g.
+    //  |-- 10 --|       |----  16 -----|    |3|      |2|     |3|           |2|
+    // "aaa 10 bbb"  =  "$'aaa ${10} bbb' - "$''" - ("10" + ("${}" * 1)) + "10"
+    //
+    //  |--- 10 --|      |------ 19 -------|     |3|       |2|    |3|       |3|     |3|        |2|    |3|
+    // "hi 10, 200"  =  "$'hi ${10}, ${200}'" - "$''" - (("10" + "${}") + ("200" + "${}")) + ("10" + "200")
+    //
+    //  |--- 10 --|      |----- 16 -----|     |3|      |1|   |1|      |3|     |3|        |2|    |3|
+    // "hi 10, 200"  =  "$'hi $x, ${200}'" - "$''" - (("x" + "$") + ("200" + "${}")) + ("10" + "200")
+    // (if x = 10)
+    //
+    // and
+    //
+    //  |3|       |--- 11 --|     |3|      | 5 |      |3|           |3|
+    // "vim"  =  "$'${"vim"}'" - "$''" - ("'vim'" + ("${}" * 1)) + "vim"
+    return template_length - QUOTES_LENGTH - (all_embedded_size + all_embedding_size) + all_stringified_size;
+}
+
+/*
+ * Apply MB_PTR_ADV() n-th
+ */
+    static void
+mb_ptr_adv_nth(char_u** p, size_t n)
+{
+    size_t	i;
+    for (i = 0; i < n; ++i)
+    {
+	MB_PTR_ADV(*p);
+    }
+}
+
+/*
+ * Replace all "${...}" of `dest` by each actual values.
+ * `source` is a template string.
+ * By `read_embedded()` or else, `source` must be checked that have no unterminated embedding.
+ */
+    static void
+embed_embedded(int (*is_closing_quote)(char_u*), char_u** dest, char_u* source, embedded_t** embedded, size_t embedded_size)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // a number of "${"
+    const size_t	VAR_EMBEDDING_START_LENGTH = 1;  // a number of head "$"
+    char_u*		to_dest;
+    char_u*		to_source = source + 2;  // Skip the first "$'"
+    size_t		finished = 0;  // a count of finished embedding
+
+    for (to_dest = *dest; !(is_closing_quote(to_source) || *to_source == NUL); MB_PTR_ADV(to_source), MB_PTR_ADV(to_dest))
+    {
+	if (to_source[0] == '$' && to_source[1] == '$')  // escaped '$'
+	{
+	    *to_dest = '$';
+	    MB_PTR_ADV(to_source);
+	    continue;
+	}
+	else if (to_source[0] == '$')  // Do embed to $var
+	{
+	    size_t	embedding_start_length = embedded[finished]->is_embedded_literal
+		? LITERAL_EMBEDDING_START_LENGTH
+		: VAR_EMBEDDING_START_LENGTH;
+	    size_t	embedding_close_length = embedded[finished]->is_embedded_literal
+		? 1
+		: 0;
+
+	    memcpy(to_dest, embedded[finished]->stringified, STRLEN(embedded[finished]->stringified));
+	    mb_ptr_adv_nth(&to_source, embedding_start_length + STRLEN(embedded[finished]->expr) + embedding_close_length - 1);  // 1 is increased by for's effect later!
+	    mb_ptr_adv_nth(&to_dest, STRLEN(embedded[finished]->stringified) - 1);  // same
+	    ++finished;
+	    continue;
+	} else if (MB_PTR2LEN(to_source) > 1) {
```

Comment:
`else` should be the next line.
`{` should be the next of the `else` line.

Reply:
:+1: